File: tools/pageAnalyzer.js
const axios = require('axios');
const cheerio = require('cheerio');
const beautify = require('js-beautify').html;

async function analyze(url) {
    try {
        const report = {
            vulnerabilities: [],
            data: ""
        };

        // 1. Основной запрос для получения кода и заголовков
        const response = await axios.get(url, {
            timeout: 10000,
            headers: { 'User-Agent': 'CyberDoc-Scanner/2.0' }
        });

        const $ = cheerio.load(response.data);
        const headers = response.headers;

        // --- АКТИВНЫЙ ТЕСТ: Проверка на Open Redirect ---
        try {
            const redirectTest = await axios.get(`${url}/?url=https://evil.com`, { maxRedirects: 0, validateStatus: null });
            if (redirectTest.status === 301 || redirectTest.status === 302) {
                report.vulnerabilities.push("[HIGH] Потенциальный Open Redirect: сервер обрабатывает внешние редиректы через параметры.");
            }
        } catch (e) {}

        // --- АКТИВНЫЙ ТЕСТ: Проверка на XSS в параметрах (Фаззинг) ---
        const xssPayload = "<script>alert(1)</script>";
        try {
            const xssTest = await axios.get(`${url}/?search=${encodeURIComponent(xssPayload)}`);
            if (xssTest.data.includes(xssPayload)) {
                report.vulnerabilities.push("[CRITICAL] Отраженная XSS: полезная нагрузка возвращается в теле ответа без фильтрации.");
            }
        } catch (e) {}

        // --- ПАССИВНЫЙ АНАЛИЗ DOM ---
        
        // Поиск форм без CSRF-защиты
        $('form').each((i, el) => {
            const hasCsrf = $(el).find('input[name*="csrf"], input[name*="token"]').length > 0;
            if (!hasCsrf && $(el).attr('method')?.toUpperCase() === 'POST') {
                report.vulnerabilities.push(`[MEDIUM] Форма обнаружена без видимого CSRF-токена: ${$(el).attr('action') || 'current page'}`);
            }
        });

        // Проверка заголовков (Безопасность транспорта)
        if (!headers['strict-transport-security']) {
            report.vulnerabilities.push("[LOW] Отсутствует HSTS заголовок: риск перехвата трафика (MITM).");
        }
        
        if (headers['server']) {
            report.vulnerabilities.push(`[INFO] Сервер раскрывает свою версию: ${headers['server']}`);
        }

        // Поиск чувствительных файлов в путях (Active Path Discovery)
        const commonPaths = ['/.env', '/.git/config', '/wp-config.php'];
        for (const path of commonPaths) {
            try {
                const check = await axios.head(`${new URL(url).origin}${path}`, { validateStatus: null });
                if (check.status === 200) {
                    report.vulnerabilities.push(`[CRITICAL] Чувствительный файл доступен: ${path}`);
                }
            } catch (e) {}
        }

        // Форматирование исходного кода
        report.data = beautify(response.data, { indent_size: 4 });

        return {
            success: true,
            vulnerabilities: report.vulnerabilities,
            data: report.data
        };

    } catch (error) {
        throw new Error(`Ошибка при активном сканировании: ${error.message}`);
    }
}

module.exports = { analyze };

File: backend/app/services/page_analyzer.py
import httpx
from bs4 import BeautifulSoup
from urllib.parse import urljoin, urlparse


class PageAnalyzer:
    """Analyses a web page for common security vulnerabilities."""

    SENSITIVE_PATHS = ("/.env", "/.git/config", "/wp-config.php", "/backup.sql")
    USER_AGENT = "CyberDoc-Scanner/3.0"

    async def analyze(self, url: str) -> dict:
        vulnerabilities: list[str] = []
        timeout = httpx.Timeout(20.0, connect=10.0)

        try:
            async with httpx.AsyncClient(
                timeout=timeout,
                headers={"User-Agent": self.USER_AGENT},
                follow_redirects=True,
                verify=False,
            ) as client:
                resp = await client.get(url)
                html = resp.text
                headers = resp.headers

                soup = BeautifulSoup(html, "lxml")

                await self._check_open_redirect(client, url, vulnerabilities)
                await self._check_reflected_xss(client, url, vulnerabilities)
                self._check_csrf_tokens(soup, vulnerabilities)
                self._check_security_headers(headers, vulnerabilities)
                await self._check_sensitive_files(client, url, vulnerabilities)

                formatted = soup.prettify()
        except httpx.TimeoutException:
            raise RuntimeError(
                f"Превышено время ожидания при подключении к {url}. "
                "Проверьте доступность ресурса."
            )
        except httpx.ConnectError as exc:
            raise RuntimeError(f"Не удалось подключиться к {url}: {exc}")

        return {
            "success": True,
            "vulnerabilities": vulnerabilities,
            "source_code": formatted,
        }

    async def _check_open_redirect(
        self, client: httpx.AsyncClient, url: str, vulns: list[str]
    ) -> None:
        try:
            resp = await client.get(
                f"{url}/?url=https://evil.com", follow_redirects=False
            )
            if resp.status_code in (301, 302):
                vulns.append(
                    "[HIGH] Потенциальный Open Redirect: "
                    "сервер обрабатывает внешние редиректы через параметры."
                )
        except Exception:
            pass

    async def _check_reflected_xss(
        self, client: httpx.AsyncClient, url: str, vulns: list[str]
    ) -> None:
        payload = "<script>alert(1)</script>"
        try:
            resp = await client.get(url, params={"search": payload})
            if payload in resp.text:
                vulns.append(
                    "[CRITICAL] Отраженная XSS: "
                    "полезная нагрузка возвращается в теле ответа без фильтрации."
                )
        except Exception:
            pass

    @staticmethod
    def _check_csrf_tokens(soup: BeautifulSoup, vulns: list[str]) -> None:
        for form in soup.find_all("form"):
            if (form.get("method") or "").upper() != "POST":
                continue
            has_token = form.find(
                "input", attrs={"name": lambda n: n and ("csrf" in n or "token" in n)}
            )
            if not has_token:
                action = form.get("action", "current page")
                vulns.append(
                    f"[MEDIUM] POST-форма без CSRF-токена: {action}"
                )

    @staticmethod
    def _check_security_headers(headers, vulns: list[str]) -> None:
        if "strict-transport-security" not in headers:
            vulns.append("[LOW] Отсутствует HSTS-заголовок: риск MITM-атаки.")
        if "x-content-type-options" not in headers:
            vulns.append("[LOW] Отсутствует X-Content-Type-Options.")
        if "x-frame-options" not in headers:
            vulns.append("[LOW] Отсутствует X-Frame-Options: риск Clickjacking.")
        server = headers.get("server")
        if server:
            vulns.append(f"[INFO] Сервер раскрывает версию: {server}")

    async def _check_sensitive_files(
        self, client: httpx.AsyncClient, url: str, vulns: list[str]
    ) -> None:
        origin = f"{urlparse(url).scheme}://{urlparse(url).netloc}"
        for path in self.SENSITIVE_PATHS:
            try:
                resp = await client.head(urljoin(origin, path))
                if resp.status_code == 200:
                    vulns.append(f"[CRITICAL] Доступен чувствительный файл: {path}")
            except Exception:
                pass
